Wrappers/Python/test/test_CIL_CVXPy.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+#   This work is part of the Core Imaging Library (CIL) developed by CCPi 
+#   (Collaborative Computational Project in Tomographic Imaging), with 
+#   substantial contributions by UKRI-STFC and University of Manchester.
+
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+
+#   http://www.apache.org/licenses/LICENSE-2.0
+
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+from cil.optimisation.functions import L2NormSquared
+from cil.optimisation.functions import TotalVariation
+
+from cil.utilities import dataexample
+import cvxpy as cp
+
+import numpy as np
+import scipy.sparse as sp
+
+import unittest
+
+class Test_CIL_vs_CVXPy(unittest.TestCase):
+
+    def setUp(self):
+
+        # default test data
+        self.data = dataexample.CAMERA.get(size=(32, 32))
+
+    
+    # Create gradient operator as sparse matrix that will be used in CVXpy
+    def sparse_gradient_matrix(self, shape, direction='forward', order=1, boundaries='Neumann', **kwargs):
+        
+        len_shape = len(shape)    
+        allMat = dict.fromkeys(range(len_shape))
+        discretization = kwargs.get('discretization',[1.0]*len_shape)
+
+        if order == 1:
+
+            for i in range(0,len_shape):
+
+                if direction == 'forward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [0,1], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[-1,:] = 0
+                    elif boundaries == 'Periodic':
+                        mat[-1,0] = 1
+
+                elif direction == 'backward':
+
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [-1,0], shape[i], shape[i], format = 'lil')
+
+                    if boundaries == 'Neumann':
+                        mat[:,-1] = 0
+                    elif boundaries == 'Periodic':
+                        mat[0,-1] = -1
+
+                tmpGrad = mat if i == 0 else sp.eye(shape[0])
```

Comment:
Could this step be explained in a comment please, or written more clearly

Reply:
Add more comments

Reply:
Thanks.  I now understand that the purpose is to create Dx and Dy. But I don't understand how this is achieved with the code lines provided. 

Dx and Dy are not explicitly mentioned, please clarify where they occur.

Please explain in detailed comments what is achieved by the two lines (79 and 83) with an embedded if-else sentence, as they are not clear to follow.

Reply:
Could you please write which lines you cannot follow? I do not have problem adding more comments but this is quite standard and a reference is provided.